gymnasium/wrappers/to_array.py

```diff
@@ -0,0 +1,244 @@
+# This wrapper will convert array inputs from an Array API compatible framework A for the actions
+# to any other Array API compatible framework B for an underlying environment that is implemented
+# in framework B, then convert the return observations from framework B back to framework A.
+#
+# More precisely, the wrapper will work for any two frameworks that can be made compatible with the
+# `array-api-compat` package.
+#
+# See https://data-apis.org/array-api/latest/ for more information on the Array API standard, and
+# https://data-apis.org/array-api-compat/ for more information on the Array API compatibility layer.
+#
+# General structure for converting between types originally copied from
+# https://github.com/google/brax/blob/9d6b7ced2a13da0d074b5e9fbd3aad8311e26997/brax/io/torch.py
+# Under the Apache 2.0 license. Copyright is held by the authors
+
+"""Helper functions and wrapper class for converting between arbitrary Array API compatible frameworks and a target framework."""
+
+from __future__ import annotations
+
+import functools
+import numbers
+from collections import abc
+from types import ModuleType
+from typing import Any, Iterable, Mapping, SupportsFloat
+
+import gymnasium as gym
+from gymnasium.core import RenderFrame, WrapperActType, WrapperObsType
+from gymnasium.error import DependencyNotInstalled
+
+
+try:
+    from array_api_compat import array_namespace, is_array_api_obj, to_device
+
+except ImportError:
+    raise DependencyNotInstalled(
+        'Array API packages are not installed therefore cannot call `to_array`, run `pip install "gymnasium[array-api]"`'
+    )
+
+
+__all__ = ["ToArray", "to_xp"]
+
+# The NoneType is not defined in Python 3.9. Remove when the minimal version is bumped to >=3.10
+_NoneType = type(None)
+Array = Any  # TODO: Switch to ArrayAPI type once https://github.com/data-apis/array-api/pull/589 is merged
+Device = Any  # TODO: Switch to ArrayAPI type if available
+
+
+def module_namespace(module: ModuleType) -> ModuleType:
+    """Determine the Array API compatible namespace of the given module.
+
+    This function is closely linked to the `array_api_compat.array_namespace` function. It returns
+    the compatible namespace for a module directly instead of from an array object of that module.
+
+    See https://github.com/data-apis/array-api-compat/blob/e14754ba0fe4c4cd51b6f45bb11a3c6609be3b5c/array_api_compat/common/_helpers.py#L442
+    """
+    if module.__name__ == "numpy":
+        from array_api_compat import numpy as numpy_namespace
+
+        return numpy_namespace
+    elif module.__name__ in ("jax.numpy", "jax"):
+        import jax.numpy
+
+        if hasattr(jax.numpy, "__array_api_version__"):
+            jp = jax.numpy
+        else:
+            import jax.experimental.array_api as jp
```

Comment:
Is there any worry that this is an experimental feature in jax

Reply:
No, on the contrary. It used to be an experimental feature, but has been stabilized and feature complete since jax v0.4.32. If we raise the minimum version, this check can be skipped and we can always use jax.numpy

Reply:
Ohh cool